bigdecimal/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+pub struct BigDecimal {
+    int_val: BigInt,
+    scale: i64,
+}
+
+impl BigDecimal {
+    /// Creates and initializes a `BigDecimal`.
+    ///
+    #[inline]
+    pub fn new(digits: BigInt, scale: i64) -> BigDecimal {
+        BigDecimal {
+            int_val: digits,
+            scale: scale,
+        }
+    }
+
+    /// Creates and initializes a `BigDecimal`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// // assert_eq!(BigDecimal::parse_bytes(b"0", 16), BigDecimal::zero());
+    /// // assert_eq!(BigDecimal::parse_bytes(b"f", 16), BigDecimal::parse_bytes(b"16", 10));
+    /// ```
+    #[inline]
+    pub fn parse_bytes(buf: &[u8], radix: u32) -> Option<BigDecimal> {
+        str::from_utf8(buf).ok().and_then(|s| BigDecimal::from_str_radix(s, radix).ok())
+    }
+
+    pub fn set_scale(&self, new_scale: i64) -> BigDecimal {
+        if self.scale == new_scale {
+            return self.clone();
+        }
+
+        if self.int_val.is_zero() {
+            return BigDecimal::new(BigInt::zero(), new_scale);
+        }
+
+        if new_scale > self.scale {
+            let raise = new_scale - self.scale;
+            // println!("raise: {}", raise);
+            let exp = (10 as i64).pow(raise as u32);
+            let new = BigDecimal::new(self.int_val.clone() * BigInt::from(exp), new_scale);
+            // println!("new: {}", new);
+            return new;
+        }
+
+        // todo implement
+        self.clone()
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ParseBigDecimalError {
+    ParseDecimal(ParseFloatError),
+    Empty,
+    Other,
+}
+
+impl fmt::Display for ParseBigDecimalError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &ParseBigDecimalError::ParseDecimal(ref e) => e.fmt(f),
+            &ParseBigDecimalError::Empty => "Failed to parse empty string".fmt(f),
+            &ParseBigDecimalError::Other => "failed to parse provided string".fmt(f),
+        }
+    }
+}
+
+impl Error for ParseBigDecimalError {
+    fn description(&self) -> &str {
+        "failed to parse bigint/biguint"
+    }
+}
+
+impl From<ParseFloatError> for ParseBigDecimalError {
+    fn from(err: ParseFloatError) -> ParseBigDecimalError {
+        ParseBigDecimalError::ParseDecimal(err)
+    }
+}
+
+impl From<ParseIntError> for ParseBigDecimalError {
+    fn from(_: ParseIntError) -> ParseBigDecimalError {
+        ParseBigDecimalError::Other
+    }
+}
+
+impl From<ParseBigIntError> for ParseBigDecimalError {
+    fn from(_: ParseBigIntError) -> ParseBigDecimalError {
+        // ParseBigDecimalError::ParseDecimal(err)
+        ParseBigDecimalError::Other
+    }
+}
+
+impl FromStr for BigDecimal {
+    type Err = ParseBigDecimalError;
+
+    #[inline]
+    fn from_str(s: &str) -> Result<BigDecimal, ParseBigDecimalError> {
+        BigDecimal::from_str_radix(s, 10)
+    }
+}
+
+impl PartialEq for BigDecimal {
+    #[inline]
+    fn eq(&self, rhs: &BigDecimal) -> bool {
```

Comment:
Java converts both arguments to BigIntegers and compares those.... Yeah that's even worse, isn't it?